lib/kube/proxy/auth.go

```diff
@@ -43,15 +43,19 @@ type kubeCreds struct {
 	kubeClient *kubernetes.Clientset
 }
 
-var skipSelfPermissionCheck bool
+type SelfPermissionsCheckTestingShimFn func(context.Context, string, authztypes.SelfSubjectAccessReviewInterface) error
```

Comment:
Sorry to jump in last minute and I realize it was implemented in a similar way before your changes as well, but this feels a little too hacky IMHO.

Replacing implementation in tests is normal but it is usually done via providing a "test" implementation of a particular interface. Looking at the code `getKubeCreds` is called from `NewForwarder` - can we add the test implementation for getting the creds to the forwarder config which tests would set? It will be cleaner and there would not be any "TestOnlyMonkeyPatchXXX"  methods in the code.

Reply:
@r0mant - So I've managed to get this working for the unit tests, and it looks OK.

It's proving _much_ harder to inject a mock for use with the integration tests, as the only point of contact between the test code and the forwarder is the initial process config - and specifying mocks for testing  in that seems like a bit overkill.

Any advice gratefully received.

Reply:
Do integration tests actually need mocking, or do they use the actual logic (as I would assume)?

Frankly, I see nothing wrong with having ability to provide custom (e.g. test) implementation even in the process config - e.g. as a part of "Proxy.Kube" service configuration. However, if the integration test needs to only skip a part of this getKubeCreds logic, you could simply add a some sort of boolean field to the config instead of a full interface and use that in the code to run/skip that code.

We already do something similar to disable various listeners in integration tests e.g. see `DisableXXX` fields on the ProxyConfig.

Reply:
The integration tests set the debug flag in the existing code, and fail if it's not set - so they definitely need some sort of switch to swap out the implementation for testing. Will follow the ProxyConfig example - thanks!

Reply:
Done